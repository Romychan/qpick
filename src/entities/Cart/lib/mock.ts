// eslint-disable-next-line @conarti/feature-sliced/layers-slices
import { Product } from '~/entities/Product/@x';

import { CartItem } from '../model/types';

/** Mock product for the cart */
export const MOCK_CART_PRODUCT: Product = {
  id: 1,
  title: 'Product',
  rate: 4.2,
  image: '/images/catalog/headphone-1.png',
  price: 1200,
  oldPrice: 1300,
  description: 'Test',
  category: 'Наушники',
};

/** List of mock cart items */
export const MOCK_CART: CartItem[] = [
  {
    product: MOCK_CART_PRODUCT,
    quantity: 2,
  },
];
