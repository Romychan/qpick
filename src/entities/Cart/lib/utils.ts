// eslint-disable-next-line @conarti/feature-sliced/layers-slices
import { Product } from '~/entities/Product/@x';

import { CartItem } from '../model/types';

/**
 * Add a product to the cart
 *
 * @param cartItems The current state of the cart
 * @param product Product to add to cart
 * @returns Updated state of the product cart
 */
export const addToCart = (cartItems: CartItem[], product: Product) => {
  const productInCart = cartItems.find(
    (item) => item.product.id === product.id,
  );

  if (!productInCart) {
    return [...cartItems, { product, quantity: 1 }];
  } else {
    return cartItems.map((item) =>
      productInCart.product.id === item.product.id
        ? { ...productInCart, quantity: item.quantity + 1 }
        : item,
    );
  }
};

/**
 * Delete a product from the cart
 *
 * @param cartItems The current state of the cart
 * @param id The ID of the product to be deleted from the cart
 * @returns Updated state of the product cart
 */
export const deleteFromCart = (cartItems: CartItem[], id: number) => {
  return cartItems.filter((item) => item.product.id !== id);
};

/**
 * Increase the number of product in the cart
 *
 * @param cartItems The current state of the cart
 * @param id Product ID to increase the quantity
 * @returns Updated state of the product cart
 */
export const increaseProductQuantity = (cartItems: CartItem[], id: number) => {
  const productInCart = cartItems.find((item) => item.product.id === id);

  if (productInCart) {
    return cartItems.map((item) =>
      productInCart.product.id === item.product.id
        ? { ...productInCart, quantity: item.quantity + 1 }
        : item,
    );
  }

  return cartItems;
};

/**
 * Decrease the number of product in the cart
 *
 * @param cartItems The current state of the cart
 * @param id Product ID to decrease the quantity
 * @returns Updated state of the product cart
 */
export const decreaseProductQuantity = (cartItems: CartItem[], id: number) => {
  const productInCart = cartItems.find((item) => item.product.id === id);

  if (productInCart) {
    return cartItems.map((item) =>
      productInCart.product.id === item.product.id
        ? { ...productInCart, quantity: item.quantity - 1 }
        : item,
    );
  }

  return cartItems;
};
