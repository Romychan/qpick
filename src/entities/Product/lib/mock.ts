import { Product } from '../model/types';

/** List of mock products */
export const MOCK_PRODUCTS: Product[] = [
  {
    id: 1,
    title: 'Apple BYZ S852I',
    rate: 4.7,
    image: '/images/catalog/headphone-1.png',
    price: 2927,
    oldPrice: 3527,
    description:
      'Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
    category: 'Наушники',
  },
  {
    id: 2,
    title: 'Apple AirPods Pro 3',
    rate: 4.5,
    image: '/images/catalog/headphone-2.png',
    price: 200,
    oldPrice: null,
    description:
      "Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.",
    category: 'Наушники',
  },
  {
    id: 3,
    title: 'Apple EarPods',
    rate: 4.5,
    image: '/images/catalog/headphone-3.png',
    price: 2327,
    oldPrice: null,
    description:
      'Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.',
    category: 'Наушники',
  },
  {
    id: 4,
    title: 'Apple BYZ S123I',
    rate: 4.7,
    image: '/images/catalog/headphone-1.png',
    price: 1500,
    oldPrice: 1950,
    description:
      'Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.',
    category: 'Наушники',
  },
  {
    id: 5,
    title: 'Apple AirPods Pro 2',
    rate: 4.5,
    image: '/images/catalog/headphone-2.png',
    price: 200,
    oldPrice: null,
    description:
      'Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.',
    category: 'Наушники',
  },
  {
    id: 6,
    title: 'Apple EarPods 2',
    rate: 3.8,
    image: '/images/catalog/headphone-3.png',
    price: 1200,
    oldPrice: 2000,
    description:
      'Aenean maximus, nisi eget mollis dictum, leo sapien sollicitudin ipsum, quis euismod ex erat ut tellus.',
    category: 'Наушники',
  },
  {
    id: 7,
    title: 'Apple AirPods',
    rate: 4.7,
    image: '/images/catalog/headphone-4.png',
    price: 9527,
    oldPrice: null,
    description:
      'Aenean maximus, nisi eget mollis dictum, leo sapien sollicitudin ipsum, quis euismod ex erat ut tellus.',
    category: 'Беспроводные наушники',
  },
  {
    id: 8,
    title: 'GERLAX GH-04',
    rate: 4.7,
    image: '/images/catalog/headphone-5.png',
    price: 6527,
    oldPrice: null,
    description:
      'Sed quis sem vitae justo feugiat tincidunt. Maecenas ut leo nisl. Praesent finibus ultricies ex et tristique.',
    category: 'Беспроводные наушники',
  },
  {
    id: 9,
    title: 'BOROFONE BO4',
    rate: 4.7,
    image: '/images/catalog/headphone-6.png',
    price: 7527,
    oldPrice: null,
    description:
      "Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.",
    category: 'Беспроводные наушники',
  },
];
