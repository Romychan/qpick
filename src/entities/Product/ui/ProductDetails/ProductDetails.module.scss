@import '~/app/assets/styles/_typography.scss';

.container {
  display: flex;
  align-items: flex-start;
  gap: 4rem;
}

.image {
  position: relative;
  width: 45%;

  & img {
    height: 100%;
    width: 100%;
    object-fit: contain;
    object-position: center;
  }
}

.content {
  display: flex;
  flex-direction: column;
  gap: 1.5rem;
  max-width: 55%;
}

.information {
}

.name {
  @include text-lg;
  @include text-semibold;
  color: var(--gray-800);
  margin-bottom: 0.25rem;
}

.wrapper {
  position: relative;
  display: flex;
  align-items: center;
  gap: 0.5rem;
}

.price {
  @include text-lg;
  @include text-semibold;
  color: var(--yellow-500);
  white-space: nowrap;
}

.oldPrice {
  @include text-sm;
  @include text-semibold;
  text-decoration: line-through;
  color: var(--yellow-300);
  white-space: nowrap;
}

.actions {
  margin-top: 1rem;
  max-width: 220px;
}

.specification {
}

.subtitle {
  @include text-sm;
  color: var(--gray-600);
}

.text {
  @include text-sm;
  color: var(--gray-900);
}

@media screen and (max-width: 576px) {
  .container {
    flex-direction: column;
  }

  .image {
    align-self: center;
  }

  .content {
    width: 100%;
    max-width: 100%;
  }
}
