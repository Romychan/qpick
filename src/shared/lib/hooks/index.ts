export { useEventListener } from './useEventListener';
export { useIsMounted } from './useIsMounted';
