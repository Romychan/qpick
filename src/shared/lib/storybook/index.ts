export { withReactHookFormDecorator } from './decorators/withReactHookFormDecorator';
export { withHeightDecorator } from './decorators/withHeightDecorator';
export { withGapDecorator } from './decorators/withGapDecorator';
export { withWidthDecorator } from './decorators/withWidthDecorator';
export { withRouterDecorator } from './decorators/withRouterDecorator';
