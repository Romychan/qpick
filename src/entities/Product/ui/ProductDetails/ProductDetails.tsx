import { ReactNode } from 'react';

import { formatPrice } from '~/shared/lib/utils';
import { Rating } from '~/shared/ui/Rating';
import { Price } from '~/shared/ui/Price';

import { Product } from '../../model/types';

import styles from './ProductDetails.module.scss';

interface IProductDetailsProps {
  /** A product for displaying detailed information */
  product: Product;
  /** The slot of the component for interactive elements */
  actions?: ReactNode;
}

/** This component is used to display detailed product information */
export const ProductDetails = ({ product, actions }: IProductDetailsProps) => {
  const { title, rate, description, price, oldPrice, image } = product;

  return (
    <div className={styles.container} data-testid="product-details">
      <div className={styles.image}>
        <img alt={title} src={image} width={250} height={270} />
      </div>

      <div className={styles.content}>
        <div className={styles.information}>
          <h1 className={styles.name}>{title}</h1>
          <Rating value={rate} />
        </div>

        <div className={styles.information}>
          <Price
            price={formatPrice(price)}
            oldPrice={formatPrice(oldPrice)}
            size="lg"
          />

          <div className={styles.actions}>{actions}</div>
        </div>

        <div className={styles.information}>
          <div className={styles.specification}>
            <p className={styles.subtitle}>Описание</p>
            <p className={styles.text}>{description}</p>
          </div>
        </div>
      </div>
    </div>
  );
};
