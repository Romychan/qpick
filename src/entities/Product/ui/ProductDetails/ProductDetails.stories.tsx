import type { Meta, StoryObj } from '@storybook/react';

import { Button } from '~/shared/ui/Button';
import {
  withRouterDecorator,
  withWidthDecorator,
} from '~/shared/lib/storybook';

import { MOCK_PRODUCTS } from '../../lib/mock';

import { ProductDetails } from './ProductDetails';

const meta = {
  title: 'Entities/Product/ProductDetails',
  component: ProductDetails,
  parameters: {
    layout: 'centered',
    width: '540px',
  },
  args: {
    product: MOCK_PRODUCTS[0],
  },
  decorators: [withRouterDecorator, withWidthDecorator],
  tags: ['autodocs'],
} satisfies Meta<typeof ProductDetails>;

export default meta;
type Story = StoryObj<typeof meta>;

/** This state is used to display the default */
export const Default: Story = {
  args: {
    actions: <Button>Actions Slot</Button>,
  },
};

/** Component without actions slot */
export const EmptySlots: Story = {
  args: {
    actions: null,
  },
};
