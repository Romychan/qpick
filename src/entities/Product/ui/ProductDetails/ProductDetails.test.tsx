import { RenderResult, render, screen } from '@testing-library/react';

import { MOCK_PRODUCTS } from '../../lib/mock';

import { ProductDetails } from './ProductDetails';

describe('ProductDetails', () => {
  let container: RenderResult;

  beforeEach(() => {
    container = render(
      <ProductDetails
        product={MOCK_PRODUCTS[0]}
        actions={<button>Actions</button>}
      />,
    );
  });

  it('should render correctly', () => {
    const nameElement = screen.getByText(/Apple/i);
    const ratingElement = screen.getByTestId('rating');
    const actionsElement = screen.getByRole('button');

    expect(nameElement).toBeInTheDocument();
    expect(ratingElement).toBeInTheDocument();
    expect(actionsElement).toBeInTheDocument();
  });

  it('should display the old price', () => {
    const oldPriceElement = screen.getByText(/3 527/i);

    expect(oldPriceElement).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    expect(container).toMatchSnapshot();
  });
});
