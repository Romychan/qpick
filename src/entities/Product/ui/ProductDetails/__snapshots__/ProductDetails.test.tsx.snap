// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductDetails > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_container_ff5a21"
        data-testid="product-details"
      >
        <div
          class="_image_ff5a21"
        >
          <img
            alt="Apple BYZ S852I"
            height="270"
            src="/images/catalog/headphone-1.png"
            width="250"
          />
        </div>
        <div
          class="_content_ff5a21"
        >
          <div
            class="_information_ff5a21"
          >
            <h1
              class="_name_ff5a21"
            >
              Apple BYZ S852I
            </h1>
            <div
              class="_container_861903"
              data-testid="rating"
            >
              <svg
                class="icon icon-star-filled _icon_861903 filled"
                data-testid="icon"
                height="24"
                width="24"
              >
                <use
                  xlink:href="/images/icons.svg#star-filled"
                />
              </svg>
              <p
                class="_text_861903"
              >
                4.7
              </p>
            </div>
          </div>
          <div
            class="_information_ff5a21"
          >
            <div
              class="_container_5a3e99 _horizontal_5a3e99 _lg_5a3e99"
              data-testid="price"
            >
              <span
                class="_price_5a3e99"
              >
                2 927 ₽
                 
              </span>
              <span
                class="_old_5a3e99"
              >
                3 527 ₽
                 
              </span>
            </div>
            <div
              class="_actions_ff5a21"
            >
              <button>
                Actions
              </button>
            </div>
          </div>
          <div
            class="_information_ff5a21"
          >
            <div
              class="_specification_ff5a21"
            >
              <p
                class="_subtitle_ff5a21"
              >
                Описание
              </p>
              <p
                class="_text_ff5a21"
              >
                Lorem Ipsum is simply dummy text of the printing and typesetting industry.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_container_ff5a21"
      data-testid="product-details"
    >
      <div
        class="_image_ff5a21"
      >
        <img
          alt="Apple BYZ S852I"
          height="270"
          src="/images/catalog/headphone-1.png"
          width="250"
        />
      </div>
      <div
        class="_content_ff5a21"
      >
        <div
          class="_information_ff5a21"
        >
          <h1
            class="_name_ff5a21"
          >
            Apple BYZ S852I
          </h1>
          <div
            class="_container_861903"
            data-testid="rating"
          >
            <svg
              class="icon icon-star-filled _icon_861903 filled"
              data-testid="icon"
              height="24"
              width="24"
            >
              <use
                xlink:href="/images/icons.svg#star-filled"
              />
            </svg>
            <p
              class="_text_861903"
            >
              4.7
            </p>
          </div>
        </div>
        <div
          class="_information_ff5a21"
        >
          <div
            class="_container_5a3e99 _horizontal_5a3e99 _lg_5a3e99"
            data-testid="price"
          >
            <span
              class="_price_5a3e99"
            >
              2 927 ₽
               
            </span>
            <span
              class="_old_5a3e99"
            >
              3 527 ₽
               
            </span>
          </div>
          <div
            class="_actions_ff5a21"
          >
            <button>
              Actions
            </button>
          </div>
        </div>
        <div
          class="_information_ff5a21"
        >
          <div
            class="_specification_ff5a21"
          >
            <p
              class="_subtitle_ff5a21"
            >
              Описание
            </p>
            <p
              class="_text_ff5a21"
            >
              Lorem Ipsum is simply dummy text of the printing and typesetting industry.
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
